x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingDeciderService.java

```diff
@@ -0,0 +1,565 @@
+    // we can assume (without user intervention) that these will eventually stop and free their currently occupied resources.
+    //
+    // The capacity is as follows:
+    //  tier: The sum total of the resources that will be removed
+    //  node: The largest block of memory that will be freed on a given node.
+    //      - If > 1 "batch" ml tasks are running on the same node, we sum their resources.
+    Optional<NativeMemoryCapacity> calculateFutureFreedCapacity(PersistentTasksCustomMetadata tasks, Duration jobMemoryExpiry) {
+        final List<PersistentTask<DatafeedParams>> jobsWithLookbackDatafeeds = datafeedTasks(tasks).stream()
+            .filter(t -> t.getParams().getEndTime() != null && t.getExecutorNode() != null)
+            .collect(Collectors.toList());
+        final List<PersistentTask<?>> assignedAnalyticsJobs = dataframeAnalyticsTasks(tasks).stream()
+            .filter(t -> t.getExecutorNode() != null)
+            .collect(Collectors.toList());
+
+        if (jobsWithLookbackDatafeeds.isEmpty() && assignedAnalyticsJobs.isEmpty()) {
+            return Optional.of(NativeMemoryCapacity.ZERO);
+        }
+        if (mlMemoryTracker.isRecentlyRefreshed(jobMemoryExpiry) == false) {
+            return Optional.empty();
+        }
+        // What is the largest chunk of memory that could be freed on a node in the future
+        Map<String, Long> maxNodeBytes = new HashMap<>();
+        for (PersistentTask<DatafeedParams> lookbackOnlyDf : jobsWithLookbackDatafeeds) {
+            Long jobSize = mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(lookbackOnlyDf.getParams().getJobId());
+            if (jobSize == null) {
+                return Optional.empty();
+            }
+            maxNodeBytes.compute(lookbackOnlyDf.getExecutorNode(), (_k, v) -> v == null ? jobSize : jobSize + v);
+        }
+        for (PersistentTask<?> task : assignedAnalyticsJobs) {
+            Long jobSize = mlMemoryTracker.getDataFrameAnalyticsJobMemoryRequirement(MlTasks.dataFrameAnalyticsId(task.getId()));
+            if (jobSize == null) {
+                return Optional.empty();
+            }
+            maxNodeBytes.compute(task.getExecutorNode(), (_k, v) -> v == null ? jobSize : jobSize + v);
+        }
+        return Optional.of(new NativeMemoryCapacity(
+            maxNodeBytes.values().stream().mapToLong(Long::longValue).sum(),
+            maxNodeBytes.values().stream().mapToLong(Long::longValue).max().orElse(0L)));
+    }
+
+    private AutoscalingDeciderResult buildDecisionAndRequestRefresh(MlScalingReason.Builder reasonBuilder) {
+        mlMemoryTracker.asyncRefresh();
+        return new AutoscalingDeciderResult(null, reasonBuilder.setSimpleReason(MEMORY_STALE).build());
+    }
+
+    private Long getAnalyticsMemoryRequirement(String analyticsId) {
+        return mlMemoryTracker.getDataFrameAnalyticsJobMemoryRequirement(analyticsId);
```

Comment:
Given that we check staleness at the beginning of the `scale` method I think it should impossible to return `null` here given where it's called from.  So:

1. Please add a Javadoc comment saying this method must only be called after checking the that memory tracker is recently refreshed
2. Add an assertion that the return value isn't `null` so we can catch unexpected situations in tests
3. Add a warning log so that if we return `null` in production and it causes a support case that autoscaling isn't working we have something in the log

Reply:
> Given that we check staleness at the beginning of the scale method I think it should impossible to return null here given where it's called from.

This method is also used in scale_up, which does not directly check for memory to not be stale. 

But, in one place where this method is used (right before scale down), we have recently checked. I will put an assert there.

Reply:
I will also put an assert in scale down. If the node load returns saying that memory is stale, that is also a weird scenario and we should fail as we just recently checked.